File: AvaloniaMedia/FFmpeg/Model/Packet.cs

using System;
using System.Threading;

using FFmpeg.AutoGen;

namespace AvaloniaMedia.FFmpeg.Model;

public sealed unsafe class Packet : IDisposable
{
    public AVPacket* PacketPointer { get; private set; } = null;

    public bool IsNull => PacketPointer == null;
    public int StreamIndex => PacketPointer->stream_index;
    public int Size => PacketPointer->size;
    public int Serial { get; set; } = -1;
    public double Duration { get; set; } = double.NaN;

    public Packet() {}

    public Packet(AVPacket* packet) => PacketPointer = packet;

    public static Packet CreatePacket()
    {
        var packet = ffmpeg.av_packet_alloc();
        if (packet == null)
            throw new FFmpegException("Failed to allocate packet");

        return new Packet(packet);
    }

    public static Packet MovePacket(Packet other)
    {
        var packetPtr = ffmpeg.av_packet_alloc();
        if (packetPtr == null)
        {
            throw new FFmpegException("Failed to allocate packet");
        }
        var packet = new Packet(packetPtr);
        ffmpeg.av_packet_move_ref(packetPtr, other.PacketPointer);
        return packet;
    }

    /// <summary>
    /// 减少Packet中堆数据的引用计数，该函数调用并不会释放AVPacket结构体本身
    /// </summary>
    public void Unref() => ffmpeg.av_packet_unref(PacketPointer);

    public void Release()
    {
        if (PacketPointer != null)
        {
            var ptr = PacketPointer;
            ffmpeg.av_packet_free(&ptr);
            PacketPointer = null;
        }
    }

    private bool disposed = false;

    ~Packet() => Dispose();

    public void Dispose()
    {
        if (Interlocked.Exchange(ref disposed, true) == true)
            return;

        Release();

        GC.SuppressFinalize(this);
    }
}

File: FFmpegNative/Native/Packet.cpp
#include "Packet.h"

#include <cassert>

#include "FFmpegException.h"

Packet::Packet(Packet &&other) noexcept
{
    this->raw_packet_ptr = other.raw_packet_ptr;
    other.raw_packet_ptr = nullptr;
}

Packet &Packet::operator=(Packet &&other) noexcept
{
    assert(this != &other);
    av_packet_free(&this->raw_packet_ptr);
    this->raw_packet_ptr = other.raw_packet_ptr;
    other.raw_packet_ptr = nullptr;
    return *this;
}

Packet::~Packet()
{
    av_packet_free(&this->raw_packet_ptr);
}

Packet Packet::create()
{
    auto ptr = av_packet_alloc();
    if (ptr == nullptr)
    {
        throw FFmpegException("Failed to allocate AVPacket");
    }
    Packet result;
    result.raw_packet_ptr = ptr;
    return result;
}

AVPacket *Packet::get()
{
    return this->raw_packet_ptr;
}

const AVPacket *Packet::get() const
{
    return this->raw_packet_ptr;
}

void Packet::reset(AVPacket *ptr) noexcept
{
    if (this->raw_packet_ptr != nullptr)
    {
        av_packet_free(&this->raw_packet_ptr);
    }
    this->raw_packet_ptr = ptr;
}

void Packet::unref()
{
    assert(this->raw_packet_ptr != nullptr);
    av_packet_unref(this->raw_packet_ptr);
}

bool Packet::is_null() const
{
    return this->raw_packet_ptr != nullptr;
}

int Packet::stream_index() const
{
    assert(this->raw_packet_ptr != nullptr);
    return this->raw_packet_ptr->stream_index;
}

int Packet::size() const
{
    assert(this->raw_packet_ptr != nullptr);
    return this->raw_packet_ptr->size;
}

int Packet::serial() const
{
    assert(this->raw_packet_ptr != nullptr);
    return this->packet_serial;
}

double Packet::duration() const
{
    assert(this->raw_packet_ptr != nullptr);
    return this->packet_duration;
}

void Packet::set_serial(int serial) noexcept
{
    this->packet_serial = serial;
}

void Packet::set_duration(double duration) noexcept
{
    this->packet_duration = duration;
}